servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java

```diff
@@ -71,74 +70,58 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
-                } else {
-                    subscriptionRequestQueueUpdater.accumulateAndGet(this, n,
-                            FlowControlUtils::addWithOverflowProtectionIfNotNegative);
-                }
-                if (!subscriptionLockOwnerUpdater.compareAndSet(this, null, currentThread)) {
-                    return;
-                }
-                // We previously added our n contribution to the queue, but now that we have acquired the lock
-                // we are responsible for draining the queue.
-                n = subscriptionRequestQueueUpdater.getAndSet(this, 0);
-                if (n == 0) {
-                    // It is possible that the previous consumer has released the lock, and drained the queue before we
-                    // acquired the lock and drained the queue. This means we have acquired the lock, but the queue has
-                    // already been drained to 0. We should release the lock, try to drain the queue again, and then
-                    // loop to acquire the lock if there are elements to drain.
-                    subscriptionLockOwner = null;
-                    n = subscriptionRequestQueueUpdater.getAndSet(this, 0);
-                    if (n == 0) {
-                        return;
-                    } else {
-                        continue;
-                    }
+            pendingDemand = mapInvalidRequestN(n);
+        } else {
+            for (;;) {
+                final long prevPendingDemand = pendingDemand;
+                if (prevPendingDemand < 0 ||
+                        pendingDemandUpdater.compareAndSet(this, prevPendingDemand,
+                                prevPendingDemand + min(Long.MAX_VALUE - prevPendingDemand, n))) {
+                    break;
                 }
             }
-            if (n < 0) {
-                subscription.cancel();
-                return; // Don't set subscriptionLockOwner = 0 ... we don't want to request any more!
+        }
+        Throwable delayedCause = null;
+        boolean tryAcquire;
+        do {
+            final long acquireId = tryAcquireReentrantLock(subscriptionLockUpdater, this);
+            if (acquireId == 0) {
+                break;
             }
+
             try {
-                subscription.request(n);
+                final long prevPendingDemand = pendingDemandUpdater.getAndSet(this, 0);
+                if (prevPendingDemand == CANCELLED) {
+                    subscription.cancel();
+                } else if (prevPendingDemand != 0) {
+                    subscription.request(prevPendingDemand);
+                }
+            } catch (Throwable cause) {
+                if (delayedCause == null) {
+                    delayedCause = cause;
+                }
             } finally {
-                subscriptionLockOwner = null;
+                tryAcquire = !releaseReentrantLock(subscriptionLockUpdater, acquireId, this);
```

Comment:
lets discuss above. I don't think we want todo this.

Reply:
I am still not sure about the approach here of continuing to call the `subscription` if a previous call to it threw. If we are not comfortable about poisoning the lock then an alternate is to set the `pending = CANCELLED` and hence ignoring any subsequent calls. If we do this, then the fast-path in `request(n)` can also avoid calling `drain()` from inside a `finally` block.

Reply:
> If we do this, then the fast-path in request(n) can also avoid calling drain() from inside a finally block.

We have two options for reacting to exceptions:
1. force cleanup internally to `ConcurrentSubscription`
2. allow cleanup to happen externally to `ConcurrentSubscription`

for (1) we have to catch the exception, force a call to `cancel()`, and we can optionally release the lock. This would require a `catch` in the fast path `request(n)` method, and a catch in the `drain()`, and we could release the lock outside the `finally`.

for (2) we have to catch/ignore exceptions while we hold the lock (because `cancel()` may come concurrently). this is the general pattern of the `ConcurrentUtil` locks in which you keep draining until you can release the lock, otherwise you may miss an event.

I would like to avoid forcing `cancel()` from `ConcurrentSubscription` as it is an interim layer and not the root cause of the exception. If the `ConcurrentSubscription` was removed from the scenario the caller would still have the same concerns related to dealing with exceptions.

Reply:
discussed offline https://github.com/apple/servicetalk/pull/1015#issuecomment-623791000